### PR TITLE
Enhanced report endpoint to support +gzip accept header (#1596)

### LINKED ISSUE #1596: Enhance QPC report endpoint to support application/json+gzip accept header
## Feature description:
QPC will upload tar.gz files to the insights upload service.  Additionally, json file size is impacting performance for the QPC download deployment reports.  If the server supported the option to provide reports as a zip, both subscription insights and the QPC CLI could benefit.

### Describe the solution you'd like
Add support for the accept header `application/json+gzip` to the following endpoints:
- `/api/v1/reports/{id}/deployments`
- `/api/v1/reports/{id}/details`

___

### PR DESCRIPTION
**Pull request by myersCody**
Closes #1596

### CODE REVIEW

File: docs/source/man.rst [hunk 1]
```diff
@@ -511,7 +511,7 @@ The ``qpc report details`` command generates a detailed report that contains the
 
 ``--json``
 
-  Displays the results of the report in JSON format. Mutually exclusive with the ``--csv`` option.
+  Returns the results of the report in Javascript Object Notation (JSON) format compressed into a tarball (tar.gz). Mutually exclusive with the ``--csv`` option.
```
Comment (kholdaway) [inline docs/source/man.rst]: I don't think we want to externalize this to the end user.  We should unzip it for them and put it in the file they specified.

File: docs/source/man.rst [hunk 2]
```diff
@@ -539,7 +539,7 @@ For example, the raw facts of a scan that includes both network and vcenter sour
 
 ``--json``
 
-  Displays the results of the report in JSON format. Mutually exclusive with the ``--csv`` option.
+  Displays the results of the report in TAR.GZ format. Mutually exclusive with the ``--csv`` option.
```
Comment (kholdaway) [inline docs/source/man.rst]: same.

File: docs/source/working_with_reports.rst [hunk 3]
```diff
@@ -4,7 +4,7 @@ When a scan job completes, it provides information that can then be viewed as a
...
 ^^^^^^^^^^^^^^^^^^^^^^^
-When the scan job completes, you have the capability to produce a report for that scan job. You can request a report with all the details, or facts, of the scan, or request a deployments report. The deployments report process runs steps to merge the facts found during the inspection of the various hosts that are contacted during the scan. When possible, the report process also runs steps to deduplicate redundant systems. For both types of reports, you can produce the report in JavaScript Object Notation (JSON) format or comma-separated values (CSV) format.
+When the scan job completes, you have the capability to produce a report for that scan job. You can request a report with all the details, or facts, of the scan, or request a deployments report. The deployments report process runs steps to merge the facts found during the inspection of the various hosts that are contacted during the scan. When possible, the report process also runs steps to deduplicate redundant systems. For both types of reports, you can produce the report in a comma-separated values (CSV) format or JavaScript Object Notation (JSON) format compressed into a tarball (tar.gz).
```
Comment (kholdaway) [inline docs/source/working_with_reports.rst]: same.

File: qpc/messages.py [hunk 4]
```diff
@@ -193,7 +193,8 @@
 REPORT_REPORT_ID_HELP = 'Report identifier.'
 REPORT_REPORT_IDS_HELP = 'Report identifiers.'
 REPORT_SCAN_JOB_IDS_HELP = 'Scan job identifiers.'
-REPORT_OUTPUT_JSON_HELP = 'Output as a JSON file.'
+REPORT_OUTPUT_JSON_HELP = 'Output as JSON file compressed into a tarball'\
```
Comment (kholdaway) [inline qpc/messages.py]: Still should be JSON.

File: quipucords/api/details_report/json_gzip_renderer.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        report_id = report_dict.get('report_id')
+        if report_id is None:
+            return None
+
+        details_report = DetailsReport.objects.filter(
+            report_id=report_id).first()
+        if details_report is None:
+            return None
+
+        json_file_name = 'details_' + str(report_id) + '.json'
```
Comment (myersCody) [inline quipucords/api/details_report/json_gzip_renderer.py]: Add a unique Id here for the tmp file, for both the tar & json

File: docs/source/man.rst [hunk 6]
```diff
@@ -511,7 +511,7 @@ The ``qpc report details`` command generates a detailed report that contains the
...
```
[the remark this answers is not in the document]
Comment (kholdaway) [reply docs/source/man.rst]: As discussed, this should indicate JSON is returned.  No tar.gz.

File: qpc/messages.py [hunk 7]
```diff
@@ -193,7 +193,7 @@
 REPORT_REPORT_ID_HELP = 'Report identifier.'
 REPORT_REPORT_IDS_HELP = 'Report identifiers.'
 REPORT_SCAN_JOB_IDS_HELP = 'Scan job identifiers.'
-REPORT_OUTPUT_JSON_HELP = 'Output as a JSON file.'
+REPORT_OUTPUT_JSON_HELP = 'Output as a tarball containing a JSON file.'
```
[the remark this answers is not in the document]
Comment (kholdaway) [reply qpc/messages.py]: As discussed, this should indicate JSON is returned.  No tar.gz.

File: qpc/report/details.py [hunk 8]
```diff
@@ -103,15 +101,21 @@ def _validate_args(self):
 
     def _handle_response_success(self):
         file_content = None
+        binary = False
         if self.args.output_json:
-            file_content = self.response.json()
-            file_content = pretty_print(file_content)
+            correct_file_ext = '.tar.gz'
```
[the remark this answers is not in the document]
Comment (kholdaway) [reply qpc/report/details.py]: 👍

File: qpc/tests_utilities.py [hunk 9]
```diff
@@ -11,7 +11,11 @@
 """Test utilities for the CLI module."""
 
 import contextlib
+import io
```
Comment (kholdaway) [inline qpc/tests_utilities.py]: The copyright date needs to be updated from 2017 to 2017-2018.

File: qpc/utils.py [hunk 10]
```diff
@@ -356,7 +358,23 @@ def write_file(filename, content):
...
     return result
+
+
+def extract_json_from_tar(fileobj_content):
+    """Extract json data from tar.gz bytes.
+
+    :param fileobj_content: BytesIo object with tarball of json dict
+    """
+    filename = '/tmp/cli_tmp_%s.tar.gz' % time.strftime('%Y%m%d_%H%M%S')
+    write_file(filename, fileobj_content, True)
```
Comment (kholdaway) [inline qpc/utils.py]: Do you delete this file later?  I will be running this to verify, but it appears to leave a tar.gz behind.

File: quipucords/api/common/util.py [hunk 11]
```diff
@@ -248,3 +253,20 @@ def expand_scanjob_with_times(scanjob, connect_only=False):
                 status_details[task_key] = task.status_message
 
     return job_json
+
+
+def create_tar_buffer(data_array):
```
Comment (kholdaway) [inline quipucords/api/common/util.py]: Is there already a tests_util.py?  Let's make sure and write unit tests for this.

File: quipucords/api/details_report/tests_details_report.py [hunk 12]
```diff
@@ -178,3 +181,42 @@ def test_csv_renderer(self):
         # pylint: disable=line-too-long
         expected = 'Report ID,Report Type,Report Version,Number Sources\r\n1,details,%s,1\r\n\r\n\r\nSource\r\nServer Identifier,Source Name,Source Type\r\n%s,test_source,network\r\nFacts\r\n\r\n' % (self.report_version, self.server_id)  # noqa
         self.assertEqual(csv_result, expected)
+
+    ##############################################################
+    # Test Json Gzip Render
+    ##############################################################
+    def test_json_gzip_renderer(self):
```
Comment (kholdaway) [inline quipucords/api/details_report/tests_details_report.py]: Let's add tests for report ID not found.  We should have these for all content types we return (should means ideally)

File: quipucords/api/deployments_report/tests_deployments_report.py [hunk 13]
```diff
@@ -323,3 +326,44 @@ def test_csv_renderer_only_name(self):
         # pylint: disable=line-too-long
         expected = 'Report ID,Report Type,Report Version\r\n1,deployments,%s\r\n\r\n\r\nSystem Fingerprints:\r\narchitecture,bios_uuid,cpu_core_count,cpu_count,cpu_socket_count,detection-network,detection-satellite,detection-vcenter,entitlements,infrastructure_type,ip_addresses,is_redhat,mac_addresses,name,os_name,os_release,os_version,redhat_certs,redhat_package_count,sources,subscription_manager_id,system_creation_date,system_last_checkin_date,virtualized_type,vm_cluster,vm_datacenter,vm_dns_name,vm_host,vm_host_socket_count,vm_state,vm_uuid\r\n,,,,,,,,,,,,,1.2.3.4,,,,,,,,,,,,,,,,,\r\n,,,,,,,,,,,,,1.2.3.4,,,,,,,,,,,,,,,,,\r\n,,,,,,,,,,,,,1.2.3.4,,,,,,,,,,,,,,,,,\r\n\r\n' % self.report_version  # noqa
         self.assertEqual(csv_result, expected)
+
+    ##############################################################
+    # Test Json Gzip Render
+    ##############################################################
+    def test_json_gzip_renderer(self):
```
Comment (kholdaway) [inline quipucords/api/deployments_report/tests_deployments_report.py]: Let's add tests for report ID not found. We should have these for all content types we return (should means ideally)

File: qpc/tests_utilities.py [hunk 14]
```diff
@@ -35,3 +39,17 @@ def redirect_stdout(stream):
         yield
     finally:
         sys.stdout = old_stdout
+
```
[the remark this answers is not in the document]
Comment (myersCody) [reply qpc/tests_utilities.py]: Actually this should go under `/api/common/utils.py`

File: quipucords/api/common/util.py (hunk 11 above)
Comment (myersCody) [reply quipucords/api/common/util.py]: I added test for this.

**Comment by kholdaway** [comment]
Approved when CI is happy.

**Pull request merged**